debezium/dbz#739 Document Amazon SNS sink (#7256)
debezium/debezium: documentation/modules/ROOT/pages/operations/debezium-server.adoc

Also please reach out if you have requirements for specific sinks to be supported by {prodname} Server or even would be interested in contributing the required implementation.
====

> {prodname} provides a ready-to-use application that streams change events from a source database to messaging infrastructure like Amazon Kinesis, Amazon SNS, Google Cloud Pub/Sub, Apache Pulsar, Redis (Stream), or NATS JetStream.

Reviewer: ```suggestion
{prodname} provides a ready-to-use application that streams change events from a source database to messaging infrastructure like Amazon Kinesis, Amazon  Simple Notification Service (SNS), Google Cloud Pub/Sub, Apache Pulsar, Redis (Stream), or NATS JetStream.
```

|===

==== Amazon SNS

> https://aws.amazon.com/sns/[Amazon Simple Notification Service (SNS)] is a managed pub/sub messaging system with support for standard and FIFO topics, fan-out delivery to multiple subscribers, and message filtering.

Reviewer: ```suggestion
https://aws.amazon.com/sns/[Amazon SNS] is a managed pub/sub messaging system with support for standard and FIFO topics, fan-out delivery to multiple subscribers, and message filtering.
```

> SNS exposes a set of REST APIs and provides a Java SDK that is used to implement the sink.

Reviewer: ```suggestion
SNS exposes a set of REST APIs and provides a Java SDK that it uses to implement sink connections.
```

> 

Reviewer: ```suggestion

Set the following properties to configure the sink: 
```

[cols="35%a,10%a,55%a",options="header"]
|===
|Property
|Default
|Description

|[[sns-type]]<<sns-type, `debezium.sink.type`>>

> |

Reviewer: ```suggestion
|No default value.
```

|Must be set to `sns`.

|[[sns-region]]<<sns-region, `debezium.sink.sns.region`>>

> |

Reviewer: ```suggestion
|No default value.
```

|The AWS region in which the SNS topics are provisioned (e.g. `us-east-1`, `eu-central-1`).

|[[sns-endpoint]]<<sns-endpoint, `debezium.sink.sns.endpoint`>>
|_endpoint determined by AWS SDK_
|(Optional) A custom endpoint URL for the SNS service.

> Useful for local development with tools such as https://localstack.cloud/[LocalStack] (e.g. `\http://localhost:4566`).

Reviewer: ```suggestion
Useful for local development with tools such as https://localstack.cloud/[LocalStack] (for example,  `\http://localhost:4566`).
```

|[[sns-credentials-profile]]<<sns-credentials-profile, `debezium.sink.sns.credentials.profile`>>

> |

Reviewer: ```suggestion
|No default value.
```

|(Optional) A credentials profile name used to communicate with the AWS API through the https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/credentials-profiles.html[default credential profiles file].

> If not present, the https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/credentials-chain.html[default credentials provider chain] is used.

Reviewer: ```suggestion
If you do not set this property, the https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/credentials-chain.html[default credentials provider chain] is used.
```

The chain looks for credentials in the following order: environment variables, Java system properties, web identity token credentials, default credential profiles file, Amazon ECS container credentials, and instance profile credentials.

|[[sns-topic-arn]]<<sns-topic-arn, `debezium.sink.sns.topic.arn`>>

> |

Reviewer: ```suggestion
|No default value.
```

> |(Optional) The default SNS topic ARN to which all messages are published when the resolved destination is neither a full ARN nor combined with a prefix.

Reviewer: ```suggestion
|(Optional) The default SNS topic Amazon Resource Names (ARNs) to which all messages are published when the resolved destination is neither a full ARN nor combined with a prefix.
```

If this property is set and no prefix is configured, all events are sent to this single topic.

|[[sns-topic-arn-prefix]]<<sns-topic-arn-prefix, `debezium.sink.sns.topic.arn.prefix`>>

> |

Reviewer: ```suggestion
|No default value.
```

This is useful to route change events to different SNS topics based on the event's destination name.

|[[sns-default-retries]]<<sns-default-retries, `debezium.sink.sns.default.retries`>>
|`5`
|The maximum number of retry attempts when publishing a batch of messages fails (either partially or completely).
Failed entries within a batch are retried individually until all entries succeed or the retry limit is reached.

|[[sns-fifo-message-group-id-header]]<<sns-fifo-message-group-id-header, `debezium.sink.sns.fifo.message.group.id.header`>>
|`aggregateId`
|(Optional) The name of the Kafka Connect record header from which the `MessageGroupId` is extracted for FIFO topics.
If the header is not present in the record, the event key is used.
If the event key is also `null`, the value of `debezium.sink.sns.fifo.default.group.id` is used.
This property is only relevant for FIFO topics (topic ARN ending in `.fifo`).

|[[sns-fifo-message-dedup-id-header]]<<sns-fifo-message-dedup-id-header, `debezium.sink.sns.fifo.message.dedup.id.header`>>

> |

Reviewer: ```suggestion
|No default value.
```

|(Optional) The name of the Kafka Connect record header from which the `MessageDeduplicationId` is extracted for FIFO topics.

> If not configured or the header is absent, no deduplication ID is set and SNS relies on content-based deduplication (if enabled on the topic).

Reviewer: ```suggestion
If this property is not set, or if the header is absent, no deduplication ID is set and SNS relies on content-based deduplication (if enabled on the topic).
```

This property is only relevant for FIFO topics.

|[[sns-fifo-default-group-id]]<<sns-fifo-default-group-id, `debezium.sink.sns.fifo.default.group.id`>>
|`default`
|(Optional) The fallback `MessageGroupId` used for FIFO topics when neither the configured header nor the event key provides a value.

|===

===== Topic ARN resolution

> The SNS sink resolves the target topic ARN for each event using the following strategy:

Reviewer: ```suggestion
The SNS sink uses the following strategy to resolve the target topic ARN for each event:
```

First, the destination name is determined.

> By default it is derived from the source topic (e.g. `myapp-inventory-products`).

Reviewer: ```suggestion
By default, the destionation name is derived from the source topic (for example, `myapp-inventory-products`).
```

> When using the xref:transformations/outbox-event-router.adoc[Outbox Event Router], the destination name is the value of the `aggregatetype` column (e.g. `user`, `notification`).

Reviewer: ```suggestion
When using the {link-prefix}:{link-transformations#outbox-event-router[Outbox Event Router], the destination name is the value of the `aggregatetype` column (for example. `user`, `notification`).
```

Author: @roldanbob which one is the expected destination?
1. https://debezium.io/documentation/reference/stable/transformations/outbox-event-router.html
2. https://debezium.io/documentation/reference/stable/transformations/index.html#outbox-event-router

> Then, the ARN is resolved from the destination name using the first matching rule:

Reviewer: ```suggestion
Then, the ARN is resolved from the destination name by using the first matching rule in the following list:
```

1. If the destination name starts with `arn:`, it is treated as a full ARN and used directly.
2. If `debezium.sink.sns.topic.arn.prefix` is configured, the ARN is composed as `prefix + destination name`.
3. If a `debezium.sink.sns.topic.arn` default is configured, that ARN is used.
4. Otherwise, the destination name is used as-is.

> This allows flexible routing strategies, from publishing all events to a single topic, to dynamically routing to per-table topics using a prefix.

Reviewer: ```suggestion
The preceding options provides a flexible routing strategy, ranging from publishing all events to a single topic, to dynamically routing events to specific destinations, based on the configured prefix.
```

> If you need more advanced topic name mapping, you can provide a custom implementation via the injection points described below.

Reviewer: ```suggestion
If you need more advanced topic name mapping, you can implement custom mappings via the injection points described in the following sections.
```

===== FIFO topics

SNS https://docs.aws.amazon.com/sns/latest/dg/sns-fifo-topics.html[FIFO topics] guarantee strict message ordering and exactly-once delivery.

> The SNS sink automatically detects FIFO topics when the resolved topic ARN ends with `.fifo` and populates the required `MessageGroupId` field.

Reviewer: ```suggestion
The SNS sink automatically routes FIFO topics when it detects the suffix `.fifo` in a resolved topic ARN.
It populates the required `MessageGroupId` field in each FIFO topic.
```

> For ordering, the `MessageGroupId` is resolved with the following precedence:

Reviewer: ```suggestion
When {prodname} sends FIFO event records to an SNS sink, the sending order depends on the source of the `MessageGroupId` value, based on the following rules of precedence:
```

> 1. The value of the header specified by `debezium.sink.sns.fifo.message.group.id.header` (defaults to `aggregateId`).

Reviewer: ```suggestion
1. The value of the header specified by xref:sns-fifo-message-group-id-header[`debezium.sink.sns.fifo.message.group.id.header`] (defaults to `aggregateId`).
```

2. The event key.
3. The value of `debezium.sink.sns.fifo.default.group.id` (defaults to `default`).

> For deduplication, if `debezium.sink.sns.fifo.message.dedup.id.header` is configured and the corresponding header is present, its value is used as the `MessageDeduplicationId`.

Reviewer: ```suggestion
For deduplication purposes, if xref:sns-fifo-message-dedup-id-header[`debezium.sink.sns.fifo.message.dedup.id.header`] is configured, and the corresponding header is present, the specified value is used as the `MessageDeduplicationId`.
```

Otherwise, SNS relies on https://docs.aws.amazon.com/sns/latest/dg/fifo-message-dedup.html[content-based deduplication] (which must be enabled on the topic).

===== Batching

> The SNS sink groups events by destination topic ARN and publishes them using the https://docs.aws.amazon.com/sns/latest/api/API_PublishBatch.html[SNS PublishBatch API] in batches of up to 10 messages (the SNS API limit).

Reviewer: ```suggestion
The SNS sink groups events by destination topic ARN and uses the https://docs.aws.amazon.com/sns/latest/api/API_PublishBatch.html[SNS PublishBatch API] to publish them in batches of up to 10 messages (the SNS API limit).
```

> Each individual message must not exceed 256 KB; the sink validates the payload size before publishing and raises an error if the limit is exceeded.

Reviewer: ```suggestion
Individual messages must not exceed 256 KB; the sink validates the payload size before publishing and raises an error if the limit is exceeded.
```

===== Injection points

> The SNS sink behaviour can be modified by a custom logic providing alternative implementations for specific functionalities.

Reviewer: ```suggestion
To apply custom logic to modify specific SNS sink behavior, you can use one of the following alternative implementations.
```

> When the alternative implementations are not available then the default ones are used.

Reviewer: ```suggestion
If no alternative implementation is available, the default behavior is used.
```
I think that you can probably remove this sentence.

[cols="35%a,10%a,55%a",options="header"]
|===
|Interface
|CDI classifier
|Description

|[[sns-ext-client]]<<sns-ext-client, `software.amazon.awssdk.services.sns.SnsClient`>>
|`@CustomConsumerBuilder`

> |Custom configured instance of a `SnsClient` used to publish messages to target topics.

Reviewer: ```suggestion
|Custom configured instance of a `SnsClient` that is used to publish messages to target topics.
```

|[[sns-ext-stream-name-mapper]]<<sns-ext-stream-name-mapper, `io.debezium.server.StreamNameMapper`>>

> |

Reviewer: ```suggestion
|Name of the source table.
```
Per comment on L833, is this the correct interpretation?

> |Custom implementation maps the planned destination (topic) name into a physical SNS topic name or ARN.

Reviewer: ```suggestion
|Custom implementation that maps the planned destination topic name to a physical SNS topic name or ARN.
```

> By default the same name is used.

Reviewer: Be explicit about the name that is used. In this context, does _same_ name, mean the name of the source table?

|===

===== Configuration examples

====== Standard topic with a fixed ARN

> All change events are published to a single standard SNS topic:

Reviewer: ```suggestion
The following example shows a possible configuration for publishing all change events to a single standard SNS topic:
```

[source,properties]
----
debezium.sink.type=sns
debezium.sink.sns.region=us-east-1
debezium.sink.sns.topic.arn=arn:aws:sns:us-east-1:123456789012:my-cdc-events
----

====== Per-table topics using a prefix

> Each source table receives its own SNS topic.

Reviewer: ```suggestion
The following example shows a possible configuration for publishing each source table to its own SNS topic.
```

The topic ARN is composed from the prefix and the destination name:

[source,properties]
----
debezium.sink.type=sns
debezium.sink.sns.region=eu-central-1
debezium.sink.sns.topic.arn.prefix=arn:aws:sns:eu-central-1:123456789012:
----

NOTE: SNS topic names do not support dots (`.`).

> Use the xref:transformations/topic-routing.adoc[Topic Routing] transformation to replace dots with hyphens (or another valid character) before the destination name is resolved to an ARN.

Reviewer: ```suggestion
Use the {link-prefix]:{link-transformations}#topic-routing[Topic Routing] transformation to replace dots with hyphens (or another valid character) before the destination name is resolved to an ARN.
```

Author: @roldanbob which one is the expected destination?
1. https://debezium.io/documentation/reference/stable/transformations/topic-routing.html
2. https://debezium.io/documentation/reference/stable/transformations/index.html#topic-routing

Reviewer: > @roldanbob which one is the expected destination?
> 
> 1. https://debezium.io/documentation/reference/stable/transformations/topic-routing.html
> 2. https://debezium.io/documentation/reference/stable/transformations/index.html#topic-routing

The link resolves to the first. 
The link syntax in my comment is required to resolve the link in both the community and product editions of the doc. 
In the product edition, links to content that resides outside of the current file cannot use the `xref` macro. For this reason, we use the `link-prefix` attribute, which is defined differently in the downstream (resolves to `link`) and upstream (`xref`) editions. 
Also, to ensure that links in the downstream docs resolve correctly, you must specify the anchor ID for the target topic. In this case to link to the Topic routing overview topic, we specify the id `topic-routing`.

For example, with a transformation that replaces `.` with `-`, a change event from the table `inventory.products` (with topic prefix `myapp`) is published to `arn:aws:sns:eu-central-1:123456789012:myapp-inventory-products`.

====== FIFO topic

> 

Reviewer: ```suggestion

The following example shows a possible configuration for publishing source tables to FIFO topics.
```

[source,properties]
----
debezium.sink.type=sns
debezium.sink.sns.region=us-east-1
debezium.sink.sns.topic.arn=arn:aws:sns:us-east-1:123456789012:my-topic.fifo
debezium.sink.sns.fifo.message.group.id.header=aggregateId
debezium.sink.sns.fifo.message.dedup.id.header=eventId
debezium.sink.sns.fifo.default.group.id=default-group
----

====== Local development with LocalStack

> 

Reviewer: ```suggestion

The following example shows a possible configuration for publishing each source table to a local AWS cloud stack.
```

[source,properties]
----
debezium.sink.type=sns
debezium.sink.sns.region=us-east-1
debezium.sink.sns.endpoint=http://localhost:4566
debezium.sink.sns.topic.arn=arn:aws:sns:us-east-1:000000000000:my-topic
----

====== Full example with MySQL source and Outbox Event Router

> This example configures {prodname} Server to capture changes from a MySQL outbox table and publish them to SNS.

Reviewer: ```suggestion
The following example shows how you might configure {prodname} Server to capture changes from a MySQL outbox table and publish them to SNS.
```

If the event key is also `null`, the value of `debezium.sink.sns.fifo.default.group.id` is used.
This property is only relevant for FIFO topics (topic ARN ending in `.fifo`).

|[[sns-fifo-message-dedup-id-header]]<<sns-fifo-message-dedup-id-header, `debezium.sink.sns.fifo.message.dedup.id.header`>>
|No default value.
|(Optional) The name of the Kafka Connect record header from which the `MessageDeduplicationId` is extracted for FIFO topics.
If this property is not set, or if the header is absent, no deduplication ID is set and SNS relies on content-based deduplication (if enabled on the topic).
This property is only relevant for FIFO topics.

|[[sns-fifo-default-group-id]]<<sns-fifo-default-group-id, `debezium.sink.sns.fifo.default.group.id`>>
|`default`
|(Optional) The fallback `MessageGroupId` used for FIFO topics when neither the configured header nor the event key provides a value.

|===

===== Topic ARN resolution

The SNS sink uses the following strategy to resolve the target topic ARN for each event:

First, the destination name is determined.
By default, the destination name is derived from the source topic (for example, `myapp-inventory-products`).

> When using the {link-prefix}:{link-transformations}#outbox-event-router[Outbox Event Router], the destination name is the value of the `aggregatetype` column (for example, `user`, `notification`).

Reviewer: ```suggestion
When using the {link-prefix}:{link-outbox-event-router}#outbox-event-router[Outbox Event Router], the destination name is the value of the `aggregatetype` column (for example, `user`, `notification`).
```

|`@CustomConsumerBuilder`
|Custom configured instance of a `SnsClient` that is used to publish messages to target topics.

|[[sns-ext-stream-name-mapper]]<<sns-ext-stream-name-mapper, `io.debezium.server.StreamNameMapper`>>
|No default value.
|Custom implementation that maps the planned destination (topic) name to a physical SNS topic name or ARN.
By default, the destination name is used directly as the SNS topic name.

|===

===== Configuration examples

====== Standard topic with a fixed ARN

The following example shows a possible configuration for publishing all change events to a single standard SNS topic:

[source,properties]
----
debezium.sink.type=sns
debezium.sink.sns.region=us-east-1
debezium.sink.sns.topic.arn=arn:aws:sns:us-east-1:123456789012:my-cdc-events
----

====== Per-table topics using a prefix

The following example shows a possible configuration for publishing each source table to its own SNS topic.

[source,properties]
----
debezium.sink.type=sns
debezium.sink.sns.region=eu-central-1
debezium.sink.sns.topic.arn.prefix=arn:aws:sns:eu-central-1:123456789012:
----

NOTE: SNS topic names do not support dots (`.`).

> Use the {link-prefix}:{link-transformations}#topic-routing[Topic Routing] transformation to replace dots with hyphens (or another valid character) before the destination name is resolved to an ARN.

Reviewer: ```suggestion
Use the {link-prefix}:{link-topic-routing}#topic-routing[Topic Routing] transformation to replace dots with hyphens (or another valid character) before the destination name is resolved to an ARN.
```
